// apify/dist/patched_apify_client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pushDataChargingContext = exports.USES_PUSH_DATA_INTERCEPTION = void 0;
exports.createPatchedApifyClient = createPatchedApifyClient;
/* eslint-disable max-classes-per-file */
const node_async_hooks_1 = require("node:async_hooks");
const apify_client_1 = require("apify-client");
const charging_js_1 = require("./charging.js");
exports.USES_PUSH_DATA_INTERCEPTION = Symbol('apify:uses-push-data-interception');
exports.pushDataChargingContext = new node_async_hooks_1.AsyncLocalStorage();
function createPatchedApifyClient(options, actor) {
    class PatchedDatasetClient extends apify_client_1.DatasetClient {
        normalizeItems(items) {
            if (typeof items === 'string') {
                const parsed = JSON.parse(items);
                return Array.isArray(parsed) ? parsed : [parsed];
            }
            if (Array.isArray(items)) {
                return items.flatMap((item) => typeof item === 'string'
                    ? JSON.parse(item)
                    : item);
            }
            return [items];
        }
        async pushItems(items) {
            const context = exports.pushDataChargingContext.getStore();
            // Normalize string inputs: a single JSON string may encode multiple items
            // (e.g. '[{"a":1},{"b":2}]'), which would be miscounted by the charging logic.
            // Parse strings into arrays so each logical item is counted individually.
            const normalizedItems = this.normalizeItems(items);
            const result = await (0, charging_js_1.pushDataAndCharge)({
                chargingManager: actor.getChargingManager(),
                items: normalizedItems,
                eventName: context?.eventName,
                isDefaultDataset: true,
                pushFn: async (limitedItems) => super.pushItems(JSON.stringify(limitedItems)), // stringify the items for faster validation in Apify client
            });
            if (!context)
                return;
            // For a single invocation of Dataset.pushData, there may be more than one call to DatasetClient.pushItems.
            // Aggregate `ChargeResult` objects across such calls.
            if (context.chargeResult === undefined) {
                context.chargeResult = result;
            }
            else {
                context.chargeResult = (0, charging_js_1.mergeChargeResults)(context.chargeResult, result);
            }
        }
    }
    class PatchedApifyClient extends apify_client_1.ApifyClient {
        dataset(id) {
            const isDefaultDataset = id === actor.config.get('defaultDatasetId');
            const datasetOptions = {
                id,
                baseUrl: this.baseUrl,
                publicBaseUrl: this.publicBaseUrl,
                apifyClient: this,
                httpClient: this.httpClient,
            };
            const hasDefaultDatasetItemEvent = charging_js_1.DEFAULT_DATASET_ITEM_EVENT in
                actor.getChargingManager().getPricingInfo().perEventPrices;
            if (!isDefaultDataset || !hasDefaultDatasetItemEvent) {
                return new apify_client_1.DatasetClient(datasetOptions);
            }
            const client = new PatchedDatasetClient(datasetOptions);
            Object.assign(client, {
                [exports.USES_PUSH_DATA_INTERCEPTION]: true,
            });
            return client;
        }
    }
    return new PatchedApifyClient(options);
}
